File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/k_closest_numbers_two_pointers.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <list>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

int binary_search(const vector<int> &a, int low, int high, int x) {
  while (low <= high) {
    int mid = low + (high - low) / 2;

    if (a[mid] < x) {
      low = mid + 1;
    } else if (a[mid] > x) {
      high = mid - 1;
    } else {
      return mid;
    }
  }

  if (low > 0) {
    low--;
  }

  return low;
}

// O(logn + k) / O(1)
vector<int> k_closest(const vector<int> &a, int k, int x) {
  int xpos = binary_search(a, 0, a.size() - 1, x);
  int cnt = 0;

  int left = xpos, right = xpos + 1;
  vector<int> res;
  while (cnt < k) {
    if (abs(x - a[left]) < abs(x - a[right])) {
      res.push_back(a[left]);
      left--;
    } else {
      res.push_back(a[right]);
      right++;
    }
    cnt++;
  }

  return res;
}

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k, x;
  cin >> tt;

  while (tt--) {
    cin >> n >> k >> x;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    auto res = k_closest(a, k, x);
    for (int r : res) {
      cout << r << " ";
    }
    cout << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/KClosestNumbersTwoPointers.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class KClosestNumbersTwoPointers {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();
        int x = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        var res = kClosestNumber(a, k, x);
        while (!res.isEmpty()) {
          System.out.printf("%d ", res.pollFirst());
        }
        System.out.println();
      }
    }
  }

  // O(logn + k) / O(1)
  public static Deque<Integer> kClosestNumber(int[] a, int k, int x) {
    int n = a.length;
    int pos = binarySearch(a, 0, n - 1, x);

    Deque<Integer> res = new ArrayDeque<>();
    int left = pos, right = pos + 1, cnt = 0;

    while (cnt < k) {
      if (left >= 0 && right < a.length) {
        if (Math.abs(x - a[left]) < Math.abs(x - a[right])) {
          res.addFirst(a[left]);
          left--;
        } else {
          res.addLast(a[right]);
          right++;
        }
      } else if (left >= 0) {
        res.addFirst(a[left]);
        left--;
      } else {
        res.addLast(a[right]);
        right++;
      }
      cnt++;
    }

    return res;
  }

  private static int binarySearch(int[] a, int low, int high, int k) {
    if (a[low] > k) {
      return low;
    }

    if (a[high] < k) {
      return high;
    }

    while (low < high) {
      int mid = low + (high - low) / 2;

      if (a[mid] < k) {
        low = mid + 1;
      } else if (a[mid] > k) {
        high = mid - 1;
      } else {
        return mid;
      }
    }

    return (Math.abs(k - a[low]) < Math.abs(k - a[high])) ? low : high;
  }
}